File: app/lib/core/notifications/push_service.dart
import 'dart:async';
import 'dart:io' show Platform;

import 'package:firebase_core/firebase_core.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../data/api/api_client.dart';
import '../../data/sync/device_id.dart';
import '../../data/sync/sync_providers.dart';
import '../auth/auth_providers.dart';
import '../auth/auth_service.dart';
import '../../features/library/providers/active_session_sync.dart';
import '../router/app_router.dart';

/// Handles a push that arrives while the app is backgrounded/terminated. Must be
/// a top-level, `vm:entry-point` function (it runs in its own isolate). Nothing
/// to do — the system renders the notification from its own payload.
@pragma('vm:entry-point')
Future<void> firebaseMessagingBackgroundHandler(RemoteMessage message) async {}

/// Registers this install's FCM token with the API for the signed-in user,
/// keeps it fresh, and routes notification taps. All Firebase calls are guarded
/// so the app is unaffected on web or when Firebase isn't initialised (tests).
class PushService {
  PushService(this._api, this._deviceId);

  final ApiClient _api;

  /// This install's id, sent with the token so the server can leave this device
  /// out of a fan-out to the reader's own devices. Resolved lazily (it lives in
  /// Drift) and never fatal — a null id just means the server can't exclude us.
  final Future<String?> Function() _deviceId;

  String? _token;
  bool _started = false;

  bool get _available => !kIsWeb && Firebase.apps.isNotEmpty;

  Future<void> start({
    void Function(RemoteMessage message)? onOpen,
    void Function()? onLendEvent,
    void Function()? onReadingEvent,
  }) async {
    if (_started || !_available) return;
    _started = true;
    final messaging = FirebaseMessaging.instance;
    await messaging.requestPermission();
    // Let notifications show while the app is foregrounded (iOS default hides them).
    await messaging.setForegroundNotificationPresentationOptions(
      alert: true,
      badge: true,
      sound: true,
    );

    await _acquireAndRegister(messaging);

    // Belt-and-braces: also catch the token if it only becomes available later
    // (e.g. APNs was slow past the poll window, or it rotates).
    messaging.onTokenRefresh.listen((t) {
      _token = t;
      _register(t);
    });

    if (onOpen != null) {
      // App launched from a notification while terminated:
      final initial = await messaging.getInitialMessage();
      if (initial != null) onOpen(initial);
      FirebaseMessaging.onMessageOpenedApp.listen(onOpen);
    }

    // A lending push arriving while the app is foregrounded → pull immediately so
    // a book just lent to me appears on the Borrowed shelf without waiting.
    // A reading push is the same idea for the running timer: the reader's other
    // device started or stopped a sitting, and this one should agree at once
    // rather than at the next 15-minute sync (owner request, 14 Aug 2026).
    if (onLendEvent != null || onReadingEvent != null) {
      FirebaseMessaging.onMessage.listen((m) {
        final type = m.data['type'];
        if (type == 'lend_new' || type == 'lend_returned' || type == 'lend_reminder') {
          onLendEvent?.call();
        } else if (type == 'reading_started' || type == 'reading_stopped') {
          onReadingEvent?.call();
        } else if (type == 'notes_changed') {
          // A note written on the reader's other device. Pull now rather than
          // at the next 15-minute tick — both devices are mid-sitting, which
          // is exactly when "immediately" is what the reader expects.
          onLendEvent?.call();
        }
      });
    }
  }

  Future<void> _acquireAndRegister(FirebaseMessaging messaging) async {
    try {
      // On iOS the FCM token exists only *after* the APNs token is set, which
      // arrives asynchronously (after registerForRemoteNotifications, which
      // requestPermission triggers). getToken() throws `apns-token-not-set`
      // until then, and onTokenRefresh does NOT fire for the *initial* token —
      // so a swallowed first failure loses it until reinstall. That's why iOS
      // never registered while Android (no APNs dependency) did. Poll first.
      if (Platform.isIOS) {
        var apns = await messaging.getAPNSToken();
        for (var i = 0; i < 20 && apns == null; i++) {
          await Future<void>.delayed(const Duration(seconds: 1));
          apns = await messaging.getAPNSToken();
        }
        if (apns == null) {
          if (kDebugMode) {
            debugPrint('PushService: APNs token never arrived — check the App ID '
                '"Push Notifications" capability / provisioning profile.');
          }
          return; // getToken() would just throw apns-token-not-set
        }
      }
      _token = await messaging.getToken();
    } catch (e) {
      if (kDebugMode) debugPrint('PushService: token fetch failed: $e');
      _token = null;
    }
    if (_token != null) await _register(_token!);
  }

  Future<void> _register(String token) async {
    try {
      String? deviceId;
      try {
        deviceId = await _deviceId();
      } catch (_) {
        deviceId = null; // registering the token still matters more than the id
      }
      await _api.registerDevice(
        token,
        Platform.isIOS ? 'ios' : 'android',
        deviceId: deviceId,
      );
    } catch (_) {
      // Best-effort; a token refresh or the next launch retries.
    }
  }

  /// On sign-out: drop the token server-side and locally so a shared device
  /// stops receiving this account's pushes.
  Future<void> stop() async {
    final token = _token;
    if (token != null) {
      try {
        await _api.unregisterDevice(token);
      } catch (_) {
        // ignore — a dead token is pruned on the next failed send anyway
      }
    }
    if (_available) {
      try {
        await FirebaseMessaging.instance.deleteToken();
      } catch (_) {}
    }
    _token = null;
    _started = false;
  }
}

final pushServiceProvider = Provider<PushService>(
  (ref) => PushService(
    ref.watch(apiClientProvider),
    // The install id, not the session's — this is device state and is wanted
    // before (and regardless of) whoever is signed in.
    () => getOrCreateDeviceId(ref.read(appDatabaseProvider)),
  ),
);

/// Where a tap on a push notification should land, from its data payload.
///
/// Pure and top-level so the rule is testable without Firebase — same reason
/// `readingTimerRouteFor` lives outside its listener. It exists at all because
/// the first version routed only lending pushes and sent *everything else* to
/// the connections inbox: tapping "『X』 is being timed on your other device"
/// opened a list of connection requests (owner report, 14 Aug 2026). A tap
/// belongs on the screen the notification is about, and a type with no obvious
/// home belongs on Home — never on an unrelated screen by default.
String pushTapRoute(Map<String, dynamic> data) {
  switch (data['type']) {
    case 'lend_new':
    case 'lend_returned':
    case 'lend_reminder':
      return Routes.lendingLedger;
    case 'connection_request':
    case 'connection_accepted':
      return Routes.connections;
    case 'reading_started':
      // The running sitting itself — the screen that can show and stop it.
      final entryId = (data['library_entry_id'] as String?)?.trim() ?? '';
      return entryId.isEmpty ? Routes.home : Routes.readingTimerPath(entryId);
    default:
      return Routes.home;
  }
}

/// Registers the token when a user signs in and clears it on sign-out; taps on a
/// notification open the screen the event is about. Watch this once at the app root.
final pushLifecycleProvider = Provider<void>((ref) {
  final push = ref.watch(pushServiceProvider);
  ref.listen<AsyncValue<KitabiAuthUser?>>(
    authStateProvider,
    (prev, next) {
      final was = prev?.valueOrNull;
      final now = next.valueOrNull;
      if (now != null && was == null) {
        push.start(
          onOpen: (message) {
            // Route the tap to where the event lives — via the external-nav
            // helper, so a cold-start tap survives the splash/bootstrap
            // redirect instead of being swallowed into home, and a tap on a
            // screen already open doesn't stack a second copy of it.
            final router = ref.read(routerProvider);
            final route = pushTapRoute(message.data);
            if (message.data['type'] == 'reading_started') {
              // A tapped notification is the one case where this device has
              // *not* seen the sitting yet: a background push does no work, so
              // nothing has mirrored the server's row into local state. Arrive
              // after the pull — the timer screen reads the session from local
              // state, finds none, and reads that as "stopped elsewhere", which
              // would bounce the reader straight back out of the screen they
              // just asked for. Best-effort: navigate even if the pull fails,
              // so a tap always goes somewhere.
              unawaited(ref
                  .read(activeSessionSyncProvider)
                  .pullAndApply()
                  .whenComplete(() => navigateFromExternal(router, route)));
              return;
            }
            navigateFromExternal(router, route);
          },
          onLendEvent: () => ref.read(syncTriggerProvider)(),
          // The payload carries the sitting, but the *server* is the authority
          // on what is running — so the push is only a nudge to re-read, which
          // also makes a missed or duplicated push harmless.
          onReadingEvent: () => ref.read(activeSessionSyncProvider).pullAndApply(),
        );
      } else if (now == null && was != null) {
        push.stop();
      }
    },
    fireImmediately: true,
  );
});

File: api/app/services/push_service.py
"""High-level push notifications — the bridge between app events and FCM.

Each `notify_*` is self-contained: it opens its own DB session, so it's safe to
hand to FastAPI `BackgroundTasks` (which run after the request's session has
closed). All are no-ops unless push is configured (`settings.push_enabled`), so
callers can enqueue them unconditionally.
"""

import uuid

import httpx

from app.core.config import get_settings
from app.core.db import SessionLocal
from app.models.profile import Profile
from app.services import device_service, fcm_client


def _display_name(p: Profile | None) -> str:
    if p is None:
        return "Someone"
    if p.full_name and p.full_name.strip():
        return p.full_name.strip()
    if p.username:
        return f"@{p.username}"
    return "Someone"


async def _push(
    target_id: uuid.UUID,
    title: str,
    body: str,
    data: dict[str, str],
    image: str | None = None,
    silent: bool = False,
    exclude_device_id: str | None = None,
) -> None:
    # Every entry point is supposed to be a no-op when push isn't configured
    # (rule 8), but only the actor-named ones checked — the self-pushes
    # (reading start/stop, notes) opened a DB session and looked up tokens on a
    # deployment with no Firebase credential at all. Guarding here covers all of
    # them, present and future.
    if not get_settings().push_enabled:
        return
    async with SessionLocal() as db:
        tokens = await device_service.tokens_for_user(
            db, target_id, exclude_device_id=exclude_device_id
        )
        if not tokens:
            return
        dead: list[str] = []
        async with httpx.AsyncClient(timeout=10) as client:
            for token in tokens:
                try:
                    result = await fcm_client.send(
                        client, token, title, body, data, image, silent=silent
                    )
                except Exception:  # noqa: BLE001 — a bad token must not sink the rest
                    result = fcm_client.ERROR
                if result == fcm_client.UNREGISTERED:
                    dead.append(token)
        await device_service.prune(db, dead)


async def _notify_from_actor(
    actor_id: uuid.UUID,
    target_id: uuid.UUID,
    title: str,
    body_suffix: str,
    data: dict[str, str],
    image: str | None = None,
) -> None:
    """Look up the actor's name and push `{name} {body_suffix}` to the target."""
    if not get_settings().push_enabled:
        return
    async with SessionLocal() as db:
        actor = await db.get(Profile, actor_id)
    name = _display_name(actor)
    await _push(target_id, title, f"{name} {body_suffix}", data, image)


async def notify_connection_request(actor_id: uuid.UUID, target_id: uuid.UUID) -> None:
    await _notify_from_actor(
        actor_id,
        target_id,
        title="New connection request",
        body_suffix="wants to connect on Kitabi",
        data={"type": "connection_request"},
    )


async def notify_connection_accepted(actor_id: uuid.UUID, target_id: uuid.UUID) -> None:
    await _notify_from_actor(
        actor_id,
        target_id,
        title="Connection accepted",
        body_suffix="accepted your connection on Kitabi",
        data={"type": "connection_accepted"},
    )


async def notify_book_lent(
    actor_id: uuid.UUID, target_id: uuid.UUID, book_title: str, book_cover: str | None = None
) -> None:
    """Someone lent a book to the target (a connected reader)."""
    await _notify_from_actor(
        actor_id,
        target_id,
        title="A book's on its way to you",
        body_suffix=f"lent you “{book_title}” on Kitabi",
        data={"type": "lend_new"},
        image=book_cover,
    )


async def notify_book_returned(
    actor_id: uuid.UUID, target_id: uuid.UUID, book_title: str, book_cover: str | None = None
) -> None:
    """The lender marked a loan returned."""
    await _notify_from_actor(
        actor_id,
        target_id,
        title="Loan marked returned",
        body_suffix=f"marked “{book_title}” returned",
        data={"type": "lend_returned"},
        image=book_cover,
    )


async def notify_return_reminder(
    actor_id: uuid.UUID, target_id: uuid.UUID, book_title: str, book_cover: str | None = None
) -> None:
    """The lender nudges a connected borrower to return a book."""
    await _notify_from_actor(
        actor_id,
        target_id,
        title="A gentle nudge",
        body_suffix=f"would like “{book_title}” back",
        data={"type": "lend_reminder"},
        image=book_cover,
    )


async def notify_reading_started(
    user_id: uuid.UUID,
    *,
    session_id: str,
    library_entry_id: str,
    started_at: str,
    device_id: str | None,
    book_title: str | None = None,
) -> None:
    """Tell the reader's *other* devices that a sitting is running.

    This is the one push that goes to your own account rather than to someone
    else, so it carries no actor name. It is visible rather than silent: a
    second phone that was asleep when the timer started should still be able to
    say what is running and offer to stop it.

    Which is exactly why the originating install must be left out **here**. The
    first cut sent to every token and echoed `device_id` for the app to ignore
    its own event — but a visible notification is rendered by the OS before any
    app code runs, so the device that started the sitting got a banner telling
    it a sitting was running "on your other device" (owner report, 14 Aug 2026).
    The exclusion is by device rather than by token because one install holds
    several tokens over its life; `device_id` stays in the payload for the
    foreground case, where the app does see it.
    """
    title = "Reading in progress"
    body = (
        f"“{book_title}” is being timed on your other device"
        if book_title
        else ("A reading session is running on your other device")
    )
    await _push(
        user_id,
        title,
        body,
        {
            "type": "reading_started",
            "session_id": session_id,
            "library_entry_id": library_entry_id,
            "started_at": started_at,
            "device_id": device_id or "",
        },
        exclude_device_id=device_id,
    )


async def notify_reading_stopped(
    user_id: uuid.UUID,
    *,
    session_id: str,
    device_id: str | None,
) -> None:
    """Silent — the other device takes its ongoing notification down.

    Announcing this would be a banner about something the reader just did on a
    device in their hand. The device that stopped it is excluded anyway: it has
    already taken its own notification down, so the push would be pure wake-up.
    """
    await _push(
        user_id,
        "",
        "",
        {
            "type": "reading_stopped",
            "session_id": session_id,
            "device_id": device_id or "",
        },
        silent=True,
        exclude_device_id=device_id,
    )


async def notify_notes_changed(user_id: uuid.UUID) -> None:
    """Silent — "there is something new, pull now".

    Deliberately carries no content: the sync engine is the only thing that
    moves note bodies, and a push that tried to would be a second source of
    truth for private text.
    """
    await _push(user_id, "", "", {"type": "notes_changed"}, silent=True)
